compat/fsmonitor/fsmonitor-fs-listen-win32.c

```diff
@@ -149,20 +147,30 @@ static void destroy_watch(struct one_watch *watch)
 static int start_rdcw_watch(struct fsmonitor_daemon_backend_data *data,
 			    struct one_watch *watch)
 {
+	DWORD dwNotifyFilter =
+		FILE_NOTIFY_CHANGE_FILE_NAME |
+		FILE_NOTIFY_CHANGE_DIR_NAME |
+		FILE_NOTIFY_CHANGE_ATTRIBUTES |
+		FILE_NOTIFY_CHANGE_SIZE |
+		FILE_NOTIFY_CHANGE_LAST_WRITE |
+		FILE_NOTIFY_CHANGE_CREATION;
+
 	ResetEvent(watch->hEvent);
 
 	memset(&watch->overlapped, 0, sizeof(watch->overlapped));
 	watch->overlapped.hEvent = watch->hEvent;
 
+start_watch:
 	watch->is_active = ReadDirectoryChangesW(
-		watch->hDir, watch->buffer, sizeof(watch->buffer), TRUE,
-		FILE_NOTIFY_CHANGE_FILE_NAME |
-		FILE_NOTIFY_CHANGE_DIR_NAME |
-		FILE_NOTIFY_CHANGE_ATTRIBUTES |
-		FILE_NOTIFY_CHANGE_SIZE |
-		FILE_NOTIFY_CHANGE_LAST_WRITE |
-		FILE_NOTIFY_CHANGE_CREATION,
-		&watch->count, &watch->overlapped, NULL);
+		watch->hDir, watch->buffer, watch->buf_len, TRUE,
+		dwNotifyFilter, &watch->count, &watch->overlapped, NULL);
+
+	if (!watch->is_active &&
+	    GetLastError() == ERROR_INVALID_PARAMETER &&
+	    watch->buf_len > MAX_BUF_WHEN_NETWORKED) {
+		watch->buf_len = MAX_BUF_WHEN_NETWORKED;
```

Comment:
I have a hunch that a smaller buffer may be required under different circumstances, too. This code would probably catch that, too, but the name `MAX_BUF_WHEN_NETWORKED` might be too misleading? How about `FALLBACK_MAX_BUF`?